scipy/_lib/_array_api.py

```diff
@@ -0,0 +1,239 @@
+    """
+    if not _GLOBAL_CONFIG["SCIPY_ARRAY_API"]:
+        # here we could wrap the namespace if needed
+        return array_api_compat.numpy
+
+    arrays = [array for array in arrays if array is not None]
+
+    compliance_scipy(*arrays)
+
+    return array_api_compat.array_namespace(*arrays)
+
+
+def as_xparray(
+    array, dtype=None, order=None, copy=None, *, xp=None, check_finite=True
+):
+    """Drop-in replacement for `np.asarray`.
+
+    Memory layout parameter `order` is not exposed in the Array API standard.
+    `order` is only enforced if the input array implementation
+    is NumPy based, otherwise `order` is just silently ignored.
+
+    The purpose of this helper is to make it possible to share code for data
+    container validation without memory copies for both downstream use cases.
+    """
+    if xp is None:
+        xp = array_namespace(array)
+    if xp.__name__ in {"numpy", "array_api_compat.numpy", "numpy.array_api"}:
+        # Use NumPy API to support order
+        if copy is True:
+            array = np.array(array, order=order, dtype=dtype)
+        else:
+            array = np.asarray(array, order=order, dtype=dtype)
+
+        # At this point array is a NumPy ndarray. We convert it to an array
+        # container that is consistent with the input's namespace.
+        array = xp.asarray(array)
+    else:
+        array = xp.asarray(array, dtype=dtype, copy=copy)
+
+    if check_finite:
+        _check_finite(array, xp)
+
+    return array
+
+
+def as_xparray_namespace(*arrays):
+    """Validate and convert arrays to a common namespace.
+
+    Parameters
+    ----------
+    *arrays : sequence of array_like
+        Arrays to validate and convert.
+
+    Returns
+    -------
+    *arrays : sequence of array_like
+        Validated and converted arrays to the common namespace.
+    namespace : module
+        Common namespace.
+
+    Notes
+    -----
+    This function is meant to be called from each public function in a SciPy
+    submodule it does the following:
+
+    1. Check for the global switch: SCIPY_ARRAY_API. This can also be accessed
+       dynamically through ``_GLOBAL_CONFIG['SCIPY_ARRAY_API']``.
+    2. `compliance_scipy` raise exceptions on known-bad subclasses. See
+       it's definition for more details.
+    3. Determine the namespace, without doing any coercion of array(-like)
+       inputs.
+    4. Call `xp.asarray` on all array.
+
+    Examples
+    --------
+    >>> import numpy as np
+    >>> x, y, xp = as_xparray_namespace(np.array([0, 1, 2]), np.array([0, 1, 2]))
+    >>> xp.__name__
+    'array_api_compat.numpy'
+    >>> x, y
+    (array([0, 1, 2]), array([0, 1, 2]))
+
+    """
+    arrays = list(arrays)
+    xp = array_namespace(*arrays)
+
+    for i, array in enumerate(arrays):
+        arrays[i] = xp.asarray(array)
+
+    return *arrays, xp
+
+
+def to_numpy(array, xp):
+    """Convert `array` into a NumPy ndarray on the CPU.
+
+    ONLY FOR TESTING
+    """
```

Comment:
Why an extra `xp` argument, is there something which cannot be done by usual `isinstance(array, torch.Tensor)` etc?